taco: init at 02-08-2022_unstable

sheepforce opened the pull request:
###### Description of changes

This includes Taco, the tensor algebra compiler. It provides sparse Einstein summation as C++ and python library and can generate self-contained C code for a given Einstein sum.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

sheepforce commented:
I am not sure what to do with the ofborg eval error, `nix build .#python3.pkgs.taco .#taco` evaluates without errors? :thinking:

markuskowa commented on pkgs/development/libraries/taco/default.nix:
```diff
@@ -0,0 +1,51 @@
+
+in stdenv.mkDerivation rec {
+  pname = "taco";
+  version = "02-08-2022_unstable";
```
```suggestion
  version = "unstable-2022-08-02";
```
(according to https://nixos.org/manual/nixpkgs/stable/#sec-package-naming)

markuskowa commented on pkgs/top-level/python-packages.nix:
```diff
@@ -11799,6 +11799,11 @@ self: super: with self; {
 
   tabview = callPackage ../development/python-modules/tabview { };
 
+  taco = toPythonModule (callPackage ../development/libraries/taco {
+    inherit python3;
```
```suggestion
  taco = toPythonModule (pkgs.taco.override {
    inherit (self) python;
```
Does this work? That construction is widely used in `python-packages.nix`.

markuskowa commented on pkgs/development/libraries/taco/default.nix:
```diff
@@ -0,0 +1,51 @@
+, fetchgit
+, cmake
+, enablePython ? false
+, python
```
Yes, usually `null` is an antipattern, but I think here it is ok. The problem is caused by the normal package, not the python package. Since the normal package lives outside the python subset setting it to null here seems to be necessary. I don't fully understand why this is the case. It seems to be a preferred solution by looking at similar packages.

sheepforce replied:
OK, I've pushed it. Let's see if ofborg is happy.

markuskowa commented on pkgs/development/libraries/taco/default.nix:
```diff
@@ -0,0 +1,51 @@
+  patches = [ ./taco.patch ];
+
+  nativeBuildInputs = [ cmake ];
+
```
```suggestion
 buildInputs = lib.optional stdenv.isDarwin llvmPackages.openmp;
```
(or simply disable darwin if it does not work).

sheepforce replied:
I've added it.

sheepforce committed:
taco: init at 02-08-2022_unstable


taco: change python to python3 


taco: fix python call


taco: pass python as null default


taco: add llvmPackages for darwin
